File: nar/src/main/java/nars/concept/snapshot/TaskLinkSnapshot.java
package nars.concept.snapshot;

import jcog.decide.Roulette;
import jcog.pri.HashedPLink;
import jcog.pri.PLink;
import jcog.pri.ScalarValue;
import jcog.pri.bag.impl.PLinkArrayBag;
import jcog.pri.op.PriMerge;
import nars.link.TaskLink;
import nars.term.Term;
import org.jetbrains.annotations.Nullable;

import java.util.Random;
import java.util.function.Predicate;

/**
 * caches an array of tasklinks tangent to an atom
 */
public final class TaskLinkSnapshot {

	public final PLinkArrayBag<Term> links = new PLinkArrayBag<>(PriMerge.replace, 0);

	public TaskLinkSnapshot() {
	}


	/**
	 * caches an AtomLinks instance in the Concept's meta table, attached by a SoftReference
	 */

	protected static int cap(int bagSize) {
		return Math.max(4, (int) Math.ceil(1f * Math.sqrt(bagSize)) /* estimate */);
		//return bagSize;
	}

	public void commit(Term x, Iterable<TaskLink> items, int itemCount, boolean reverse) {

		links.capacity(cap(itemCount));

		links.commit();

		int xh = x.hashCodeShort();
		for (TaskLink t : items) {
			Term y = t.other(x, xh, reverse);
			if (y != null)
				links.put(new HashedPLink<>(y, t.pri()));
		}
	}


	public @Nullable Term sample(Predicate<Term> filter, byte punc, Random rng) {
		@Nullable Object[] ll = links.items();
		int lls = Math.min(links.size(), ll.length);
		if (lls == 0)
			return null;
		else {

            int li = Roulette.selectRouletteCached(lls, (int i) -> {

				PLink<Term> x = (PLink) ll[i];
				return x != null && filter.test(x.id) ?
					Math.max(ScalarValue.EPSILON,
					//x.priPunc(punc)
					x.pri()
					)
					:
					Float.NaN;

			}, rng::nextFloat);

            PLink<Term> l = li >= 0 ? (PLink<Term>) ll[li] : null;

			return l != null ? l.id : null;
		}

	}


	public final boolean isEmpty() {
		return links.isEmpty();
	}
}

File: nar/src/main/java/nars/concept/snapshot/TaskLinkSnapshot.kt
package nars.concept.snapshot

import jcog.decide.Roulette
import jcog.math.FloatSupplier
import jcog.pri.HashedPLink
import jcog.pri.PLink
import jcog.pri.PriReference
import jcog.pri.ScalarValue
import jcog.pri.bag.impl.PLinkArrayBag
import jcog.pri.op.PriMerge
import nars.link.TaskLink
import nars.term.Term
import org.eclipse.collections.api.block.function.primitive.IntToFloatFunction
import java.util.*
import java.util.function.Predicate

/**
 * caches an array of tasklinks tangent to an atom
 */
class TaskLinkSnapshot {

    val links = PLinkArrayBag<Term>(PriMerge.replace, 0)


    val isEmpty: Boolean
        get() = links.isEmpty

    fun commit(x: Term, items: Iterable<TaskLink>, itemCount: Int, reverse: Boolean) {

        links.capacity<jcog.pri.bag.Bag<Term, PriReference<Term>>>(cap(itemCount))

        links.commit()

        val xh = x.hashCodeShort()
        for (t in items) {
            val y = t.other(x, xh, reverse)
            if (y != null)
                links.put(HashedPLink(y, t.pri()))
        }
    }


    fun sample(filter: Predicate<Term>, punc: Byte, rng: Random): Term? {
        val ll = links.items()
        val lls = Math.min(links.size(), ll!!.size)
        if (lls == 0)
            return null
        else {

            val li = Roulette.selectRouletteCached(lls, IntToFloatFunction { i ->
                val x = ll[i] as PLink<*>
                if (x != null && filter.test(x.id as Term))
                    Math.max(ScalarValue.EPSILON,
                            //x.priPunc(punc)
                            x.pri()
                    )
                else
                    java.lang.Float.NaN
            }, FloatSupplier { rng.nextFloat() })

            val l = if (li >= 0) ll[li] as PLink<Term> else null

            return l?.id
        }

    }

    companion object {


        /**
         * caches an AtomLinks instance in the Concept's meta table, attached by a SoftReference
         */

        protected fun cap(bagSize: Int): Int {
            return Math.max(4, Math.ceil(1 * Math.sqrt(bagSize.toDouble())).toInt() /* estimate */)
            //return bagSize;
        }
    }
}
